Unify with Codeplay Graph extension. (#4)
reble/llvm: sycl/doc/extensions/proposed/sycl_ext_oneapi_graph.asciidoc

graph ahead of execution.

The following benefits would become possible if the user could define a
dependency graph to the SYCL runtime prior to execution:

* Reduction in runtime overhead by only submitting a single graph object, rather
  than many individual commands.

* Enable more work to be done offline, in particular producing a graph ahead of
  time allows for improved performance at runtime from reduced overhead.

* Unlock DMA hardware features through graph analysis by the runtime.

* Whole graph optimizations become available, including but not limited to:
** Kernel fusion/fission.
** Inter-node memory reuse from data staying resident on device.
** Identification of the peak intermediate output memory requirement, used for
   more optimal memory allocation.

As well as benefits to the SYCL runtime, there are also advantages to the user
developing SYCL applications, as repetitive workloads no longer have to
redundantly issue the same sequence of commands. Instead, a graph is only
constructed once and submitted for execution as many times as is necessary, only
changing the data in input buffers or USM allocations. For machine learning

> applications where the same command group pattern is run repeatedly for

Reviewer: ```suggestion
changing the data in input buffers or USM allocations. For applications from specific domains, such as machine learning, 
where the same command group pattern is run repeatedly for
```

additive changes to the API, and so in introduced in future versions of the
extension.

Another reason for deferring a serialize/deserialize API (8) is that its scope
could extend from emitting the graph in a binary format, to emitting a
standardized IR format that enables further device specific graph optimizations.

Multi-device support (7) is something we are looking into introducing into
the extension, which may result in API changes.

=== Graph Building Mechanisms

This extension contains two different API mechanisms for constructing a graph
of commands:

1. **Explicit graph building API** - Allows users to specify the exact nodes
and edges they want to add to the graph.

2. **Queue recording API (aka "Record & Replay")** - Introduces state to a
`sycl::queue` such that rather than scheduling commands immediately for
execution, they are added to the graph object instead, with edges based on the
data dependencies of the command group.

Each of these mechanisms for constructing a graph have their own advantages, so
having both APIs available allows the user to pick the one which is most
suitable for them. The queue recording API allows quicker porting of existing

> applications, and can capture work done by a library in the graph. While the

Reviewer: ```suggestion
applications, and can capture external work that is submitted to a queue, for example via library function calls. While the
```

redundantly issue the same sequence of commands. Instead, a graph is only
constructed once and submitted for execution as many times as is necessary, only
changing the data in input buffers or USM allocations. For machine learning
applications where the same command group pattern is run repeatedly for
different inputs, this is particularly useful.

=== Requirements

In order to achieve the goals described in previous sections, the following
requirements were considered:

1. Ability to update inputs/outputs of the graph between submissions, without
   changing the overall graph structure.
2. Enable low effort porting of existing applications to use the extension.
3. Profiling, debugging, and tracing functionality at the granularity of graph
   nodes.
4. Integrate sub-graphs (previously constructed graphs) when constructing a new
   graph.
5. Support the USM model of memory as well as buffer model.
6. Compatible with other SYCL extensions and features, e.g kernel fusion &
   built-in kernels.
7. Ability to record a graph with commands submitted to different devices in the
   same context.
8. A graph constructed using a device queue may be executed on another compatible

>    queue.

Author: We think this could be difficult with the current DPC++ prototype, so should discuss if this is a requirement or not

9. Capability to serialize graphs to a binary format which can then be
   de-serialized and executed. This is helpful for offline cases where a graph
   can be created by an offline tool to be loaded and run without the end-user
   incurring the overheads of graph creation.
10. Backend interoperability, the ability to retrieve a native graph object from
    the graph and use that in a native backend API.

To allow for prototype implementations of this extension to be developed
quickly for evaluation the scope of this proposal was limited to a subset
of these requirements. In particular, the serialization functionality (9),
backend interoperability (10), and a profiling/debugging interface (3) were
omitted. As these are not easy to abstract over a number of backends without
significant investigation. It is also hoped these features can be exposed as
additive changes to the API, and so in introduced in future versions of the
extension.

Another reason for deferring a serialize/deserialize API (9) is that its scope
could extend from emitting the graph in a binary format, to emitting a
standardized IR format that enables further device specific graph optimizations.

> Multi-device support (7) is something we are looking into introducing into

Author: We should discuss the requirement of allowing an executable graph to be submitted which contains commands targetting different devices. This is not currently possible, but is a feature request we've had from colleagues in Codeplay.

Author: Now tracked in https://github.com/reble/llvm/issues/7

== Introduction

> Through the use of command groups SYCL is already able to create a DAG of kernel

Reviewer 2: ```suggestion
Through the use of command groups SYCL is already able to create a dependency graph (in form of a directed acyclic graph) of kernel
```

   more optimal memory allocation.

As well as benefits to the SYCL runtime, there are also advantages to the user
developing SYCL applications, as repetitive workloads no longer have to
redundantly issue the same sequence of commands. Instead, a graph is only
constructed once and submitted for execution as many times as is necessary, only
changing the data in input buffers or USM allocations. For applications from
specific domains, such as machine learning, where the same command group pattern
is run repeatedly for different inputs, this is particularly useful.

=== Requirements

In order to achieve the goals described in previous sections, the following
requirements were considered:

1. Ability to update inputs/outputs of the graph between submissions, without
   changing the overall graph structure.
2. Enable low effort porting of existing applications to use the extension.
3. Profiling, debugging, and tracing functionality at the granularity of graph
   nodes.
4. Integrate sub-graphs (previously constructed graphs) when constructing a new
   graph.
5. Support the USM model of memory as well as buffer model.

> 6. Compatible with other SYCL extensions and features, e.g kernel fusion &

Reviewer 2: ```suggestion
6. Compatible with other SYCL extensions and features, e.g. kernel fusion &
```

Reviewer 2: I guess it makes sense to re-read the whole draft before submitting it to improve its language.

Author: Updated this, the main problem I have with re-reading the whole draft is that I've spent so long working on this document that I find it quite hard to read objectively without projecting what I think it says. Also, as a native English speaker I've never had to learn grammar rules properly 😄 

If you have time to do that pass Julian it'd be great, otherwise for now I'd tend towards relying on external reviewers coming to the spec for the first time commenting when the language isn't clear to them.

   built-in kernels.
7. Ability to record a graph with commands submitted to different devices in the
   same context.
8. Capability to serialize graphs to a binary format which can then be
   de-serialized and executed. This is helpful for offline cases where a graph
   can be created by an offline tool to be loaded and run without the end-user
   incurring the overheads of graph creation.
9. Backend interoperability, the ability to retrieve a native graph object from
    the graph and use that in a native backend API.

To allow for prototype implementations of this extension to be developed
quickly for evaluation the scope of this proposal was limited to a subset
of these requirements. In particular, the serialization functionality (8),
backend interoperability (9), and a profiling/debugging interface (3) were
omitted. As these are not easy to abstract over a number of backends without
significant investigation. It is also hoped these features can be exposed as

> additive changes to the API, and so in introduced in future versions of the

Reviewer 2: ```suggestion
additive changes to the API, and thus introduced in future versions of the
```

Each of these mechanisms for constructing a graph have their own advantages, so
having both APIs available allows the user to pick the one which is most
suitable for them. The queue recording API allows quicker porting of existing
applications, and can capture external work that is submitted to a queue, for
example via library function calls. While the explicit API can better express
what data is internal to the graph for optimization, and dependencies don't need
to be inferred.

It is valid to combine these two mechanisms sequentially when constructing a

> graph, however it is not valid to concurrently use them. An error will be thrown

Reviewer 3: ```suggestion
graph, however it is not valid to use them concurrently. An error will be thrown
```
Very minor suggestion I just think it reads better.

|1     |Initial extension version. Base features are supported.
|===

=== SYCL Graph Terminology

:explicit-memory-ops: https://www.khronos.org/registry/SYCL/specs/sycl-2020/html/sycl-2020.html#subsec:explicitmemory

Table 2. Terminology.
[%header,cols="1,3"]
|===
| Concept | Description

| Graph
| `command_graph` class that stores structured commands and their dependencies.

A SYCL graph is a collection of commands (nodes) and their dependencies (edges).
From the SYCL perspective, this graph will be acyclic and directed (DAG) as
users cannot express a cycle in the core SYCL API.

| Node
| A command, which can have different attributes.

> When recording a queue to construct a graph, nodes in a SYCL graph represent

Reviewer: ```suggestion
Nodes in a SYCL `command_graph` represent
```

Reviewer: I think its better to keep the first part of the description generic. And describe restrictions for the record&replay mode (forward edges only) and/or explicit mode in a second part (subsections?)

Author: I've split this up into a table with the generic description, followed by subsections for each of the modes with their own table for how nodes & edges are realized. Which I agree looks cleaner rather than mixing these all in the same description.

Author: One thing I've said is that in the explicit API subsection is that edges can be created from buffer accessors even if the user doesn't specify any node dependencies. Could you confirm if this is the case or not?

e.g
```cpp

 sycl::buffer buf;
 auto n1 = graph.add(
        [&](sycl::handler& cgh){
        auto acc = buf.get_access<access::mode::read_write>(cgh);
         cgh.parallel_for(sycl::range<1>{n},
		  [=](item<1> id){acc[id]=1; }
		 );
    });

auto n2 = graph.add(
        [&](sycl::handler& cgh){
	 auto acc = buf.get_access<access::mode::read_write>(cgh);
         cgh.parallel_for(sycl::range<1>{n},
		  [=](item<1> id){acc[id] *= 2; }
		 );
    });

// Does n2 depend on n1 because of the buffer accessor?

```

Reviewer: Yes that should be supported.

Author: cool, makes sense but wanted to check 👍

a data dependency between two nodes. These dependencies are expressed by the user
code through buffer accessors. There is also the partial ability to track USM
data dependencies provided the pointers used in the graph nodes are the same.
With the limitation that a node taking an offset USM pointer input will not be
identified as having an edge to another node taking a pointer input to the base
address of the same USM allocation.

In the explicit graph building API, `make_edge()` is used to define the dependency
rather than inferring them from data dependencies.
|===

=== API Modifications

[source, c++]
----
namespace sycl {
namespace ext::oneapi::experimental {

// State of a queue, returned by info::queue::state
enum class queue_state {
  executing,
  recording
};

class node {};

void make_edge(node sender, node receiver);

// State of a graph
enum class graph_state {
  modifiable,
  executable
};

// New object representing graph
template<graph_state State = graph_state::modifiable>
class command_graph {};

template<>
class command_graph<graph_state::modifiable> {
public:
  command_graph(const property_list &propList = {});

>   command_graph<graph_state::executable> finalize(context &syclContext) const;

Author: We need to decide if `syclContext` should be passed by value or const reference instead

Reviewer 3: We've updated this to be a const ref here. Given our previous discussions about issues with the plain reference, it makes sense in this context to make it const. It's also consistent with most other usages of the context in the SYCL spec.

  node add(const std::vector<node>& dep = {});

  template<typename T>
  node add(T cgf, const std::vector<node>& dep = {});

  node add_malloc_device(void *&data, size_t numBytes, const std::vector<node>& dep = {});
  node add_free(void *data, const std::vector<node>& dep = {});
};

template<>
class command_graph<graph_state::executable> {
public:
    command_graph() = delete;

>     void update(const command_graph<graph_state::modifiable> &graph);

Author: We included the executable graph update feature for `SYCL_EXT_CODEPLAY_GRAPHS` but we should work out if there are ML usecases (e.g a pytorch workload) which justifies the inclusion of
this feature in the extension.

Author: Discussion of this feature now tracked in https://github.com/reble/llvm/issues/8

};
}  // namespace ext::oneapi::experimental

// New methods added to the sycl::queue class
using namespace ext::oneapi::experimental;
class queue {
public:
  bool begin_recording(command_graph<graph_state::modifiable> &graph);
  bool end_recording();
  event submit(command_graph<graph_state::executable> graph);

> };

Reviewer 3: As previously discussed we could switch this submit call to accept a handler in line with regular queue submits, that would enable us to use `depends_on` to schedule the graph with other queue submissions, like so:
```c++
auto ev = q.submit([&](handler& cgh){
  cgh.depends_on(some_other_event);
  //Not sure of the most appropriate name here
  cgh.exec_graph(g);
});
```
I think this makes sense in keeping with exisiting submission semantics, and it shouldn't present any implementation difficulties. I am keen to hear others thoughts and suggestions though!

Author: Agree with this, is more symmetric with the rest of the SYCL API. We could also add a queue convenience shortcut function for this. Don't have a better suggestion for naming than `exec_graph` though.

Author: Moved this discussion to https://github.com/reble/llvm/issues/21 and can make a standalone follow-up PR for it

}  // namespace sycl
----

=== Node

Node is a class that encapsulates tasks like SYCL kernel functions or host tasks
for deferred execution. A graph has to be created first, the structure of a
graph is defined second by adding nodes and edges.

[source,c++]
----
namespace sycl::ext::oneapi::experimental {
  class node {};
}
----

=== Edge

A dependency between two nodes representing a happens-before relationship.
`sender` and `receiver` may be associated to different graphs.

[source,c++]
----
namespace sycl::ext::oneapi::experimental {
  void make_edge(node sender, node receiver); // Adds a dependency between two nodes
}
----

Parameters:

* `sender` - Node which will be a dependency of `receiver`.

* `receiver` - Node which will be dependent on `sender`.

Exceptions:

> * TODO - Throw if this introduces a cycle?

Author: Do we want `make_edge`  to throw an exception if it is used to introduce a cycle to the graph?

Reviewer: This could make the function `make_edge` implicitly become a costly operation especially for graphs at large scale. Because you wold have to run the analysis each time you connect two nodes. Another option would be throwing when the executable graph is created. And/or keeping it optional.

Author: Having this error when the finalized graph is created instead makes sense to avoid repeatedly paying the cost of analysing the graph. Seems like the kind of check that would be useful for users while developing an application, but once they actually want to release the application would be unwanted overhead.

So if there's a mechanism for making this optional, that could be good. CUDA has [cudaGraphInstantiateWithFlags](https://docs.nvidia.com/cuda/cuda-runtime-api/group__CUDART__GRAPH.html#group__CUDART__GRAPH_1ga2c652a24ba93e52b99a47bec0888233),
we could potentially add a flags parameter to `finalize()` that would allow a user to pass in configuration information for creating the executable graph? E.g we could have a `validate` flag that would do this cycle check and possibly other verifications. Other optimization flags could be added here in the future if deemed necessary.

Author: Moved this discussion to https://github.com/reble/llvm/issues/12, and will remove this TODO

these can be seen as specialized kernels executing on the device.

| Edge
| An edge in a queue recorded graph represents a data dependency between two
nodes. Data dependencies can naturally be expressed in user code through buffer
accessors. USM pointers also convey data dependencies, however offsets into
system allocations (`malloc`/`new`) are not supported.
|===

=== API Modifications

[source, c++]
----
namespace sycl {
namespace ext::oneapi::experimental {

// State of a queue, returned by info::queue::state
enum class queue_state {
  executing,
  recording
};

class node {};

// State of a graph
enum class graph_state {
  modifiable,
  executable
};

// New object representing graph
template<graph_state State = graph_state::modifiable>
class command_graph {};

template<>
class command_graph<graph_state::modifiable> {
public:
  command_graph(const property_list &propList = {});
  command_graph<graph_state::executable> finalize(const context &syclContext) const;

  node add(const std::vector<node>& dep = {});

  template<typename T>
  node add(T cgf, const std::vector<node>& dep = {});

  node add_malloc_device(void *&data, size_t numBytes, const std::vector<node>& dep = {});

>   node add_free(void *data, const std::vector<node>& dep = {});

Reviewer 2: Should we use add_malloc and add_free to be consistent?

Author: We discussed this on the call, and `add_malloc_device` was chosen as it's consistent with the USM `malloc_device` API. Which got me thinking about the templated and not templated versions of that function and whether we should do that too. So I created https://github.com/reble/llvm/issues/25 to track that, and we can continue any discussion on the naming of the function there too if you like.

on things that have already happened). This means that transparently a node can
depend on a previously recorded graph (sub-graph), which works by creating edges
to the individual nodes in the old graph. Explicit memory operations without
kernels, such as a memory copy, are still classed as nodes under this
definition, as the {explicit-memory-ops}[SYCL 2020 specification states] that
these can be seen as specialized kernels executing on the device.

| Edge
| An edge in a queue recorded graph represents a data dependency between two
nodes. These dependencies are expressed by the user code through buffer
accessors. There is also the partial ability to track USM data dependencies
provided the pointers used in the graph nodes are the same. With the limitation
that a node taking an offset USM pointer input will not be identified as having
an edge to another node taking a pointer input to the base address of the same
USM allocation.
|===

=== API Modifications

[source, c++]
----
namespace sycl {
namespace ext::oneapi::experimental {

// State of a queue, returned by info::queue::state
enum class queue_state {
  executing,
  recording
};

class node {};

> void make_edge(node sender, node receiver);

Reviewer: ```suggestion
```

// State of a graph
enum class graph_state {
  modifiable,
  executable
};

// New object representing graph
template<graph_state State = graph_state::modifiable>
class command_graph {};

template<>
class command_graph<graph_state::modifiable> {
public:
  command_graph(const property_list &propList = {});
  command_graph<graph_state::executable> finalize(const context &syclContext) const;

> 

Reviewer: ```suggestion
  void make_edge(node sender, node receiver);
```

Reviewer: We recently changed `make_edge` from free standing function to member function of a modifiable graph. To avoid cross graph synchronization.

Author: Ahh see that in the PoC branch, updated

};
}  // namespace sycl
----

=== Node

Node is a class that encapsulates tasks like SYCL kernel functions or host tasks
for deferred execution. A graph has to be created first, the structure of a
graph is defined second by adding nodes and edges.

[source,c++]
----
namespace sycl::ext::oneapi::experimental {
  class node {};
}
----

=== Edge

A dependency between two nodes representing a happens-before relationship.
`sender` and `receiver` may be associated to different graphs.

[source,c++]
----
namespace sycl::ext::oneapi::experimental {
  void make_edge(node sender, node receiver); // Adds a dependency between two nodes
}
----

Parameters:

* `sender` - Node which will be a dependency of `receiver`.

* `receiver` - Node which will be dependent on `sender`.

Exceptions:

* Throws synchronously with error code `invalid` if a queue is recording
  commands to any graph associated with `sender` or `receiver`.

=== Graph

:crs: https://www.khronos.org/registry/SYCL/specs/sycl-2020/html/sycl-2020.html#sec:reference-semantics

This extension adds a new `command_graph` object which follows the
{crs}[common reference semantics] of other SYCL runtime objects.

> `command_graph` is a class that represents a directed acyclic graph of nodes. A
> graph can have different states, can be nested, can have multiple root nodes
> that are scheduled for execution first and multiple leaf nodes that are
> scheduled for execution last. The execution of a graph has been completed when
> all leaf node tasks have been completed.

Reviewer: ```suggestion
A `command_graph` represents a directed acyclic graph of nodes, where each node represents a single command or a sub-graph. The execution of a graph completes when all of its nodes have completed.
```

submissions of the program. Each submission encompasses either one or both of
a.) some data movement, b.) a single asynchronous kernel launch. Nodes cannot
define forward edges, only backwards (i.e. kernels can only create dependencies
on things that have already happened). This means that transparently a node can
depend on a previously recorded graph (sub-graph), which works by creating edges
to the individual nodes in the old graph. Explicit memory operations without
kernels, such as a memory copy, are still classed as nodes under this
definition, as the {explicit-memory-ops}[SYCL 2020 specification states] that
these can be seen as specialized kernels executing on the device.

| Edge
| An edge in a queue recorded graph represents a data dependency between two
nodes. These dependencies are expressed by the user code through buffer
accessors. There is also the partial ability to track USM data dependencies
provided the pointers used in the graph nodes are the same. With the limitation
that a node taking an offset USM pointer input will not be identified as having
an edge to another node taking a pointer input to the base address of the same

> USM allocation.

Reviewer 3: I wonder about including this wording because in some ways it seems like a implementation detail or limitation. In theory I think a runtime should be able to track usm allocations, though I'm not sure if DPC++ does that currently so it may be a general limitation in practice.

Author: That's a good point, this wording is based on a limitation in the ComputeCpp prototype which I don't think is a general limitation. I think if we devoted engineering resources to looking into this, we'd be able to track the USM data dependencies better in an implementation for anything that's not `usm_system_allocations`. I'll try to rephrase this wording.

is building the structure of a graph first and adding tasks later.

Parameters:

* `dep` - Nodes the created node will be dependent on.

Returns: The empty node which has been added to the graph.

Exceptions:

* Throws synchronously with error code `invalid` if a queue is recording
  commands to the graph.

|
[source,c++]
----
template<typename T>
node add(T cgf, const std::vector<node>& dep = {});
----
|This function adds a command group function object to a graph. The function
object can contain single or multiple commands such as a host task which is
scheduled by the SYCL runtime or a SYCL function for invoking kernels with all
restrictions that apply as described in the core specification.

Parameters:

* `cgf` - Command group function object to be added as a node

* `dep` - Nodes the created node will be dependent on.

Returns: The command-group function object node which has been added to the graph.

Exceptions:

* Throws synchronously with error code `invalid` if a queue is recording
  commands to the graph.

|
[source,c++]
----
void make_edge(node sender, node receiver);
----

|Creates a dependency between two nodes representing a happens-before relationship.

> Preconditions:

Reviewer 2: Should we add this precondition to the add() and memory routines as well?

Author: good shout, done that

----
|Adding a node that encapsulates a `malloc` operation.

Parameters:

* `data` - Return parameter set to the address of memory allocated.

* `numBytes` - Size in bytes to allocate.

* `dep` - Nodes the created node will be dependent on.

Returns: The memory allocation node which has been added to the graph

|
[source,c++]
----
node add_free(void *data, const std::vector<node>& dep = {});
----
|Adding a node that encapsulates a `free` operation.

Parameters:

* `data` - Address of memory to free.

* `dep` - Nodes the created node will be dependent on.

Returns: The memory freeing node which has been added to the graph.

Exceptions:

* TODO - Throw if not allocated by `add_malloc_device`?
* TODO - Throw if already freed?

> * TODO - Throw if not valid address?

Author: Should `add_free` error under any of these circumstances?

Author: Tracked in https://github.com/reble/llvm/issues/10, and will remove the TODOs for this PR

----
using namespace ext::oneapi::experimental;
bool queue::begin_recording(command_graph<graph_state::modifiable> &graph)
----

|Synchronously changes the state of the queue to the `queue_state::recording`
state.

Parameters:

* `graph` - Graph object to start recording commands to.

Returns: `true` if the queue was previously in the `queue_state::executing`
state, `false` otherwise.

Exceptions:

* Throws synchronously with error code `invalid` if the queue is already
  recording to a different graph.

|
[source, c++]
----
bool queue::end_recording()
----

|Synchronously changes the state of the queue to the `queue_state::executing`
state.

Returns: `true` if the queue was previously in the `queue_state::recording`
state, `false` otherwise.

|
[source,c++]
----
using namespace ext::oneapi::experimental;
event queue::submit(command_graph<graph_state::executable> graph)
----

|When invoked with the queue in the `queue_state::recording` state, a graph is
added as a subgraph node. When invoked with the queue in the default
`queue_state::executing` state, the graph is submitted for execution. Support
for submitting a graph for execution, before a previous execution has been

> completed is backend specific. The runtime may throw an error.

Author: "Support for submitting a graph for execution, before a previous execution has been
completed is backend specific. The runtime may throw an error."

I used the original SYCL_EXT_ONEAPI_GRAPH wording here, but SYCL_EXT_CODEPLAY_GRAPHS says

"There are no gurantees that more than one instance of `graph` will execute concurrently.
 Submitting a graph for execution, before a previous execution has been completed
 may result in serialized execution depending on the SYCL backend and
 characteristics of the graph."

We should discuss what we want to do here.

Author: Now tracked in https://github.com/reble/llvm/issues/9

When a queue is in recording mode asynchronous exceptions will not be
generated, as no device execution is occuring. Synchronous errors specified as
being thrown in the default queue executing state, will still be thrown when a
queue is in the recording state.

The `queue::begin_recording` and `queue::end_recording` entry-points return a
`bool` value informing the user whether a state change occurred. False is
returned rather than throwing an exception when state isn't changed. This design
is because the queue is already in the state the user desires, so if the
function threw an exception in this case, the application would likely swallow
it and then proceed.

While a queue is in the recording state, methods performed on that queue which
are not command submissions behave as normal. This includes waits, throws, and
queries on the queue. These are all ignored by the graph system, as opposed to
throwing an exception when in queue recording mode. This is because otherwise
there would be no thread safe way for a user to check they could call these
functions without throwing, as a query about the state of the queue may be
immediately stale.

> * TODO - error on add_node while being recorded to a queue? or queue recording a

Author: I haven't added any specification wording on what happens a user mixes graph building mechanisms.

If a modifiable graph is being recorded to by a queue, and then the user uses explicit APIs to add nodes to the same graph before the recording has finished, is that okay? We should specify if that's valid or not.

Another possible scenario is if a modifiable graph contains nodes which have been explicitly added, and then the user goes and records more commands to it from a queue. Is that okay?

Author: Now tracked in https://github.com/reble/llvm/issues/11, and i'll remove the TODO

----
|Adding a node that encapsulates a `free` operation.

Parameters:

* `data` - Address of memory to free.

* `dep` - Nodes the created node will be dependent on.

Returns: The memory freeing node which has been added to the graph.

Exceptions:

* Throws synchronously with error code `invalid` if a queue is recording
  commands to the graph.

|===

Table 8. Member functions of the `command_graph` class (executable graph update).
[cols="2a,a"]
|===
|Member function|Description

|
[source, c++]
----
void command_graph<graph_state::executable> update(const command_graph<graph_state::modifiable> &graph);
----

|Updates the executable graph node inputs & outputs from a topologically
identical modifiable graph. The effects of the update will be visible
on the next submission of the executable graph without the need for additional
user synchronization.

Parameters:

* `graph` - Modifiable graph object to update graph node inputs & outputs with.
  This graph must have the same topology as the original graph used on
  executable graph creation.

Preconditions:

* This member function is only available when the `command_graph` state is
  `graph_state::executable`.

Exceptions:

> * Throws synchronously with error code `invalid` if the topology of `graph` is

Reviewer 2: Do we need more specific errors? A whole-graph update seems error-prone and hard to debug to me.

Author: When we looked through the SYCL spec the wording around what's thrown on errors tends to be relatively sparse and limited to error codes. My understanding is that the runtime should put more helpful error information in the `what()` message of the exception, but I don't think we need to specify exactly what form that returned string will take.

Author: Going to resolve this thread since the current wording is what we have in the Codeplay spec, and I don't want to expand the scope of this PR too much from just merging the two specs. However, if you want to discuss it further please make an Issue for it, because it is a good question.

   command-groups submitted to the queue.

2. `property::queue::enable_profiling` - This property has no effect on graph
   recording. When set on the queue a graph is submitted to however, it allows
   profiling information to be obtained from the event returned by a graph
   submission.

For any other queue property that is defined by an extension, it is the
responsibility of the extension to define the relationship between that queue
property and this graph extension.

==== New Queue Member Functions

Table 8. Additional member functions of the `sycl::queue` class.
[cols="2a,a"]
|===
|Member function|Description

|
[source, c++]
----
using namespace ext::oneapi::experimental;
bool queue::begin_recording(command_graph<graph_state::modifiable> &graph)
----

|Synchronously changes the state of the queue to the `queue_state::recording`
state.

Parameters:

* `graph` - Graph object to start recording commands to.

Returns: `true` if the queue was previously in the `queue_state::executing`
state, `false` otherwise.

Exceptions:

* Throws synchronously with error code `invalid` if the queue is already
  recording to a different graph.

|
[source, c++]
----

> bool queue::end_recording()

Reviewer 2: Should we add `using namespace ext::oneapi::experimental;` to all signatures to be consistent?

Author: Although this is the one signature where I think we don't need it for correctness, there are others that should have it, so included it everywhere for consistency

...
----

=== Diamond Dependency

The following snippet of code shows how a SYCL `queue` can be put into a
recording state, which allows a `command_graph` object to be populated by the
command-groups submitted to the queue. Once the graph is complete, recording
finishes on the queue to put it back into the default executing state. The
graph is then finalized so that no more nodes can be added. Lastly, the graph is
submitted as a whole for execution via
`queue::submit(command_graph<graph_state::executable>)`.

[source, c++]
----
  queue q{default_selector{}};

  // New object representing graph of command-groups
  ext::oneapi::experimental::command_graph graph;
  {
    buffer<T> bufferA{dataA.data(), range<1>{elements}};
    buffer<T> bufferB{dataB.data(), range<1>{elements}};
    buffer<T> bufferC{dataC.data(), range<1>{elements}};

    // `q` will be put in the recording state where commands are recorded to
    // `graph` rather than submitted for execution immediately.
    q.begin_recording(graph);

>     // Record commands to `graph` with the following topology.

Reviewer 2: Should we add the topology analogously to the dot product example?

Author: Could you clarify what change you'd like to see here? The samples are doing different things, so have different topologies. We could make both record & replay and explicit API have the same sample but using the two different mechanisms, although it seems like reductions are problematic in the implementation at the moment.

What I have done is add an ASCII diagram to the top of the dot product example like the one here, to illustrate the current different in topologies

Author: Going to resolve this thread, and we can do future sample updates in a later PR if desired.

// . Submit overload of a queue. `submit(graph)` Use a combination of `executable_graph::exec_and_wait()` and queue property `sycl::ext::oneapi::property::queue::lazy_execution{}` instead.
// . `class graph<graph_state>` Use dedicated `class graph` (equivalent to `graph_state == modifiable`) and `class executable_graph` (equivalent to `graph_state == executable`) instead.

== Issues

=== Multi Device Graph

Allow an executable graph to contain nodes targeting different devices.

**Outcome:** Under consideration

=== Record & Replay: Mark Internal Memory

When a graph is created by recording a queue there is no way to tag memory
objects internal to the graph, which would enable optimizations on the internal
memory. Do we need an interface record & replay can use to identify buffers and
USM allocations not used outside of the graph?

**Outcome:** Unresolved

=== Executable Graph Update

Is there a ML usecase (e.g pytorch workload) which justifies the inclusion of
this feature in the extension.

**Outcome:** Unresolved

> === Graph Submimssion Synchronization

Reviewer: ```suggestion
=== Graph Submission Synchronization
```
